## Tbx5 gene expression analysis
### Introduction

This part of the work provides the analysis that was performed using the
data from three ChIP-seq experiments which analyzed gene expression of 
*Tbx5* transcription factor in *Homo sapiens* organism.\

The experiments that were used for *Homo sapiens* *Tbx5* gene expression
analysis are listed below.\

***Homo sapiens:***\

 - [1. EXP037713 ](https://gtrd.biouml.org/#!table/gtrd_current.experiments/Details/ID=EXP037713)(*iPSCs-derived cardiomyocytes*)\
 - [2. EXP053380 ](https://gtrd.biouml.org/#!table/gtrd_current.experiments/Details/ID=EXP053380)(*iPSCs-derived cardiomyocytes (Treatment: day 30)*)\
 - [3. EXP037721 ](https://gtrd.biouml.org/#!table/gtrd_current.experiments/Details/ID=EXP037721)(*iPSCs-derived cardiomyocytes (Treatment: G296S mutant)*)\

The analysis provides the following information:\
 **1.** Total peak counts of each performed experiment;\
 **2.** Peak count distribution within different chromosomes.\

### Total peak counts
The barplot below shows how many *Tbx5* gene peaks were detected while analyzing
cardiomyocyte cells from control sample, 30-day treatment sample and missense
mutation-affected sample. The labels that are present on ***x*** axis
are the experiment abbreviations where 'hg' means 'Homo sapiens' and a number
defines an experiment number from the list provided above.

```{r fig, fig.width = 4, fig.height = 5, fig.align = "center"}
library(rtracklayer)
library(randomcoloR)
library(ggplot2)
library(ggthemes)
library(plyranges)
library(ggpubr)

# Importing BigBed format files that were generated after
# performing three separate experiments with Homo sapiens:
hg1 <- import("../inputs/Homo_sapiens/PEAKS040223_TBX5_Q99593_GEM_1005.bb")
hg2 <- import("../inputs/Homo_sapiens/PEAKS050535_TBX5_Q99593_GEM_1005.bb")
hg3 <- import("../inputs/Homo_sapiens/PEAKS040231_TBX5_Q99593_GEM_1005.bb")

# Creating a GRanges list that will be used for further analysis:
gr_hglist <- GRangesList(hg1, hg2, hg3)

# Creating a vector that stores total peak counts:
peak_counts <- c()
peak_counts[c("hg1", "hg2", "hg3")] <- c(length(gr_hglist[[1]]), length(gr_hglist[[2]]), length(gr_hglist[[3]]))

# Plotting peak count per experiment:
object <- c("hg1", "hg2", "hg3")
exp_peaks <- c(peak_counts[1], peak_counts[2], peak_counts[3])
col_fill <- randomColor(1, hue = "orange", luminosity="dark")

dfpeak_count <- data.frame(object, exp_peaks)
bar_plot <- ggplot(dfpeak_count, aes(x=object, y=exp_peaks)) +
    geom_bar(stat="identity", position="dodge", width=0.4,
             color="black", fill=col_fill) +
    geom_text(aes(label=exp_peaks), position=position_dodge(width=0.9),
              vjust=-0.25) +
    labs(title="Peak counts within experiments", x="Experiment",
         y="Peak counts") +
    theme(plot.title = element_text(hjust = 0.5))

bar_plot + theme(
    panel.background = element_rect(fill = "#eeeef1", colour="#4c0001"),
    panel.grid.major.y = element_line(colour="#cab5b5", size=0.3, linetype="dashed"),
    panel.grid.minor.y = element_line(colour="#cab5b5", size=0.3, linetype="dashed"),
    panel.grid.major.x = element_line(colour="#cab5b5", size=0.2, linetype="longdash"),
    panel.grid.minor.x = element_line(colour="#cab5b5", size=0.2, linetype="longdash")
)

# Determining which experiment has a maximum peak count:
max_count <- dfpeak_count[which.max(dfpeak_count$exp_peaks),]
min_count <- dfpeak_count[which.min(dfpeak_count$exp_peaks),]
```
According to the barchart, the greatest number of *Tbx5* gene expression peaks was
detected in the cells that were analyzed in ***```r max_count$object```*** experiment.
The smallest number of *Tbx5* gene peaks was observed in the dataset that was retrieved
from ***```r min_count$object```*** experiment.

### Peak counts by chromosome
The barcharts below represent peak count distribution for every chromosome
within three experiments.

```{r fig2, fig2.width = 6, fig2.height = 5, fig2.align = "center"}
hg_frame1 <- as.data.frame(gr_hglist[[1]])
hg_unq1 <- unique(hg_frame1$seqnames)
hg1df <- as.data.frame(hg_unq1)
hg1_chr <- hg1df[order(as.integer(sub(".*?(\\d+)", "\\1", hg1df$hg_unq1))), ]

chr_mt <- data.matrix(hg1_chr[grep("^(chr([1-9]+|[XY])|[XY])$", hg1_chr)])
chr_c <- chr_mt[, 1]

exp1_chr_peaks <- c()
exp2_chr_peaks <- c()
exp3_chr_peaks <- c()

for (chr in chr_c) {
    peaks1 <- gr_hglist[[1]] %>% filter(seqnames == chr)
    exp1_chr_peaks <- c(exp1_chr_peaks, length(peaks1))
    peaks2 <- gr_hglist[[2]] %>% filter(seqnames == chr)
    exp2_chr_peaks <- c(exp2_chr_peaks, length(peaks2))
    peaks3 <- gr_hglist[[3]] %>% filter(seqnames == chr)
    exp3_chr_peaks <- c(exp3_chr_peaks, length(peaks3))
}

df1 <- data.frame(chr=chr_c, peaks=exp1_chr_peaks)
df2 <- data.frame(chr=chr_c, peaks=exp2_chr_peaks)
df3 <- data.frame(chr=chr_c, peaks=exp3_chr_peaks)

df1$chr <- factor(df1$chr, levels = df1$chr)
df2$chr <- factor(df2$chr, levels = df2$chr)
df3$chr <- factor(df3$chr, levels = df3$chr)


plot1 <- ggplot(df1, aes(x=chr, y=peaks)) +
            ylim(0, 400) +
            geom_bar(stat = "identity", color="#498350", fill="#6bc175") +
            geom_text(aes(label=peaks, vjust=-1)) +
            labs(title="Experiment EXP037713", x="Chromosomes",y="Peak count") +
            theme(plot.title = element_text(hjust = 0.5), panel.background = element_rect(fill = "#e4ffef"))

plot2 <- ggplot(df2, aes(x=chr, y=peaks)) +
            ylim(0, 1750) +
            geom_bar(stat = "identity", color="#6b3a3b", fill="#b56364") +
            geom_text(aes(label=peaks, vjust=-1)) +
            labs(title="Experiment EXP053380", x="Chromosomes",y="Peak count") +
            theme(plot.title = element_text(hjust = 0.5), panel.background = element_rect(fill = "#ffcace"))

plot3 <- ggplot(df3, aes(x=chr, y=peaks)) +
            ylim(0, 1750) +
            geom_bar(stat = "identity", color="#054f6f", fill="#0777a7") +
            geom_text(aes(label=peaks, vjust=-1)) +
            labs(title="Experiment EXP037721", x="Chromosomes",y="Peak count") +
            theme(plot.title = element_text(hjust = 0.5), panel.background = element_rect(fill = "#d7ebff"))


ggarrange(plot1, plot2, plot3, labels = c("A", "B", "C"), ncol = 1, nrow=3)
```
