## Tbx5 transcription factor analysis
### Introduction

This part of the work provides the analysis that was performed using the
data from seven ChIP-seq experiments which analyzed Tbx5 transcription factor
abundance in *Mus musculus* organism.\

The experiments that were used for *Mus musculus* Tbx5 transcription factor
analysis are listed below.\

***Mus musculus:***\

 - [1. EXP058852 ](https://gtrd.biouml.org/#!table/gtrd_current.experiments/
                    Details/ID=EXP058852)(*heart ventricle*)\
 - [2. EXP030898 ](https://gtrd.biouml.org/#!table/gtrd_current.experiments/
                    Details/ID=EXP030898)(*HL-1 (cardiac muscle cells)*)\
 - [3. EXP062056 ](https://gtrd.biouml.org/#!table/gtrd_current.experiments/
                    Details/ID=EXP062056)(*neonatal cardiac fibroblast cells
                                          line expressing large T antigen*)\
 - [4. EXP058843 ](https://gtrd.biouml.org/#!table/gtrd_current.experiments/
                    Details/ID=EXP058843)(*mouse embryonic fibroblasts*)\
 - [5. EXP058847 ](https://gtrd.biouml.org/#!table/gtrd_current.experiments/
                    Details/ID=EXP058847)(*mouse embryonic fibroblasts*)\
 - [6. EXP058850 ](https://gtrd.biouml.org/#!table/gtrd_current.experiments/
                    Details/ID=EXP058850)(*mouse embryonic fibroblasts*)\
 - [7. EXP058856 ](https://gtrd.biouml.org/#!table/gtrd_current.experiments/
                    Details/ID=EXP058856)(*mouse embryonic fibroblasts*)\

The analysis provides the following information:\
 **1.** Total peak counts of each performed experiment;\
 **2.** Peak count distribution within different chromosomes.\

```{r Libraries}
# Loading used libraries:
library(pacman)
p_load(data.table, rtracklayer, randomcoloR, ggplot2, ggthemes,
       plyranges, ggpubr, BRGenomics, reshape2, plotly, heatmaply,
       dplyr, gplots, genomation, ggseqlogo, BSgenome.Mmusculus.UCSC.mm10,
       BSgenome, Biostrings, scales)
```

```{r Paths}
# Declaring paths:
INTERMEDIATE_FILES <- "../Intermediate_data_files/"
INPUTS <- "../Inputs/"
FIGURES <- "../Figures/"
```

### Total peak counts
The barplot below shows how many peaks were detected while analyzing
*Mus musculus* heart cells that were treated with different experimental
conditions.

```{r Peaks_in_samples, fig.width = 4, fig.height = 6, fig.align = "center"}
# Creating Sample Key table:
samples <- read.csv(file = paste0(INTERMEDIATE_FILES, "sample_key.csv"))

# Importing BigBed format files:
bbfiles <- list.files(path = paste0(INPUTS, "BigBed/"), "*bb")
grl <- GRangesList()

# Declaring what chromosomes are analyzed:
chr_abr <- c(paste0("chr", 1:19), "chrX", "chrY")

# Creating GRanges objects and extracting certain chromosomes:
for (i in 1:length(bbfiles)) {
    grl[[i]] <- import(paste0(INPUTS, "BigBed/", bbfiles[i])) %>%
                       filter(seqnames %in% chr_abr)
    names(grl)[i] <- samples$Graph_names[samples$Filename == bbfiles[i]]
}

# Creating peak dataframe that keeps data about total peak counts in
# each sample:
peaks <- lapply(grl, length) %>%
            as.data.table()  %>%
            melt() %>%
            as.data.frame()  %>%
            setnames(., c("Experiment", "Peak_count"))

# Picking random dark red color:
col_fill <- randomColor(1, hue = "red", luminosity = "dark")

# Creating a barplot that visualizes peak differences in samples:
plot1 <- ggplot(peaks, aes(x = Experiment, y = Peak_count)) +
            geom_bar(stat = "identity", position = "dodge", width = 0.5,
                     color = "black", fill = "#930d1f") +
            labs(x = "", y = "Pikų skaičius", size = 5) +
            ylim(0, 200000) +
            scale_y_continuous(labels = label_number(suffix = " K",
                               scale = 1e-3)) +
            theme(panel.background = element_rect(fill = "#eeeef1",
                                                  colour = "#4c0001"),
                  panel.grid.major.y = element_line(colour = "#cab5b5",
                                           size = 0.3, linetype = "dashed"),
                  panel.grid.minor.y = element_line(colour = "#cab5b5",
                                           size = 0.3, linetype = "dashed"),
                  panel.grid.major.x = element_line(colour = "#cab5b5",
                                           size = 0.2, linetype = "longdash"),
                  panel.grid.minor.x = element_line(colour = "#cab5b5",
                                           size = 0.2, linetype = "longdash"),
                  axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1,
                                   size = 11, face = "bold", color = "black"),
                  axis.text.y = element_text(size = 11, face = "bold",
                                   color = "black"),
                  axis.title.x = element_text(size = 2),
                  axis.title.y = element_text(size = 16),
                  plot.title = element_text(hjust = 0.5, face = "bold")) +
                  annotate("rect", xmin = 1.5, xmax = 4.5, ymin = -1, size = 1,
                           ymax = 170000, alpha = 0, color = "#ffae00") +
                  annotate("rect", xmin = 5.5, xmax = 6.5, ymin = -1, size = 1,
                           ymax = 25000, alpha = 0, color = "#ffae00")

# Saving generated image:
png(file = paste0(FIGURES, "total_peak_counts.png"), width = 400)
plot1
dev.off()
plot1

# Determining which experiment has a maximum peak count:
max_count <- peaks[which.max(peaks$Peak_count), ]
min_count <- peaks[which.min(peaks$Peak_count), ]
```

According to the barchart, the greatest number of peaks was determined in the
cells from ***```r max_count```*** sample.
The smallest number of peaks was observed in the dataset that was
retrieved from ***```r min_count```*** sample.

### Peak counts by chromosome
The barcharts below show how peak counts are distributed within
different chromosomes. Each bar represents separate sample.

```{r Peaks_in_chr, fig.width = 7, fig.height = 6, fig.align = "center"}
# Disabling scientific notation:
options(scipen = 100)

# Creating a dataframe that stores data about peak counts in each
# chromosome for different samples:
peak_counts <- data.frame(matrix(ncol = 3, nrow = 0)) %>%
                    setNames(., c("Name", "Chromosome", "Peak_count"))

# Defining chromosome lenghts (Mbp) based on MGI data:
chr_mbp <- c(195, 182, 160, 157, 152, 150, 145, 130, 124, 131,
             122, 120, 121, 125, 104, 98, 95, 91, 61, 169, 91)

# Calculating chromosome lenghts in bp:
chr_bp <- chr_mbp * 1000000

# Declaring a function that calculates how many peaks are in each chromosome:
count_peaks <- function(name, objects) {
    for (chr in 1:21) {
            peaks <- objects[[name]] %>%
                    filter(seqnames == chr_abr[chr]) %>%
                    length() / chr_bp[chr] * 100000000
            peak_counts[nrow(peak_counts) + 1,] = c(name, chr_abr[chr], peaks)
        }
        return(peak_counts)
    }

peak_counts <- lapply(names(grl), count_peaks, objects = grl) %>% bind_rows()

# write.csv(peak_counts, paste0(INTERMEDIATE_FILES, "peaks_by_chr_nnorm.csv"))
# write.csv(peak_counts, paste0(INTERMEDIATE_FILES, "peaks_by_chr_norm.csv"))

# Calling factor() function in order to maintain certain Chromosome and Name
# order:
cas <- unique(peak_counts$Chromosome)
peak_counts$Chromosome <- factor(peak_counts$Chromosome, levels = c(cas))
peak_counts$Name <- factor(peak_counts$Name, levels = unique(peak_counts$Name))

# Note: it is essential to add 'as.numeric'. Otherwise, Peak_count is
# considered as 'string' type - column values are represented in
# lexicographic order which leads to wrong visualization.

# Creating barplots that visualize peak differences between different
# chromosomes:
plot2 <- ggplot(peak_counts, aes(x = Name, y = as.numeric(Peak_count))) +
            geom_bar(stat = "identity", color = "black", fill = col_fill) +
            facet_wrap(~ Chromosome, ncol = 7) +
            ylab("Pikų skaičius") +
            xlab("") +
            scale_y_continuous(labels = label_number(suffix = " K",
                               scale = 1e-3)) +
            theme_linedraw() +
            theme(axis.text.x = element_text(angle = 90, size = 12,
                  vjust = 0.5),
                  legend.position = "none",
                  axis.text.y = element_text(size = 14, face = "bold"),
                  axis.title.x = element_text(size = 14, colour = "black"),
                  axis.title.y = element_text(size = 20, colour = "black"),
                  strip.background = element_rect(fill = "white"),
                  strip.text = element_text(colour = "black", face = "bold",
                                            size = 16))

# Saving generated image:
png(file = paste0(FIGURES, "peak_counts_by_chr.png"), width = 1200,
    height = 500)
plot2
dev.off()
plot2
```

### Peak overlaps between samples
The heatmap below helps to indicate sample similarity be determining
the percentage of the overlaping peak regions.

```{r Sample_similarity, fig.width = 6, fig.height = 6, fig.align = "center"}
# Creating a matrix that is filled with 'NA':
coef_matrix <- matrix(nrow = length(grl), ncol = length(grl))

# Declaring a function that calculates modified Jaccard coefficient:
jaccard <- function(a, b) {
    len <- reduce(c(grl[[a]], grl[[a]])) %>% length()
    return((length(GenomicRanges::intersect(grl[[a]], grl[[b]])) / len) * 100)
}

# Calculating Jaccard coefficient for sample pair:
for (i in 1:length(grl)) {
    for (y in 1:length(grl)) {
        coef_matrix[i, y] = jaccard(i, y)
    }
}

# Setting colnames and rownames for the matrix:
colnames(coef_matrix) <- names(grl)
rownames(coef_matrix) <- names(grl)

coef_mat1 <- coef_matrix
coef_mat2 <- coef_matrix

# Passing Jaccard coefficients to matrix except for the diagonal - it
# contains 'NA':
coef_mat1[lower.tri(coef_mat1, diag = TRUE)] <- NA
coef_mat2[upper.tri(coef_mat2, diag = TRUE)] <- NA

# Binding two matrixes:
coef_mat_bind <- rbind(data.matrix(coef_mat1), data.matrix(coef_mat2))

# Translating matrix to dataframe using melt() function:
melt_coef_mat <- melt(coef_mat_bind, na.rm = TRUE)

# Creating a heatmap that shows similarity between samples:
plot3 <- ggplot(melt_coef_mat, aes(x = Var2, y = Var1, fill = value)) +
            geom_tile(color = "black") +
            geom_text(aes(label = round(value, digits = 3)), size = 4.5,
                          color = "#030101") +
            labs(x = "", y = "") +
            scale_fill_gradient(low = "#ffee8e", high = "#ab1f1f") +
            guides(fill = guide_colourbar(title = "Koeficientas",
                   face = "bold")) +
            theme(axis.text = element_text(size = 12, colour = "black",
                                           face = "bold"),
                  axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
                  axis.title.x = element_text(size = 14, colour = "black"),
                  axis.title.y = element_text(size = 14, colour = "black"),
                  panel.grid.major = element_line(color = "#eeeeee"),
                  plot.title = element_text(hjust = 0.5, size = 16,
                                            face = "bold"),
                  legend.position = "bottom")

# Saving generated image:
png(file = paste0(FIGURES, "peak_overlaps.png"), width = 450)
plot3
dev.off()
plot3
```

### Motif determination using PWM matrix
The following step of analysis determines Tbx5 motif hit count in FASTA files
that store sequences of the corresponding genomic positions using positional
weight matrix (PWM) of Tbx5 transcription factor.

```{r Tbx5_motif_counts, fig.width = 5, fig.height = 6, fig.align = "center"}
# Reading Tbx5 transcription factor PWM matrix:
mpwm <- read.table(paste0(INTERMEDIATE_FILES, "TBX5_MOUSE.H11MO.0.D.pwm"))
mpwm <- t(mpwm)

# Setting matrix rownames:
rownames(mpwm) <- c("A", "C", "G", "T")

# Generating Tbx5 sequence logo and saving the image:
tbx5_motif <- ggseqlogo(mpwm)
png(file = paste0(FIGURES, "tbx5_motif.png"), height = 320)
tbx5_motif
dev.off()

# Calculating Tbx5 hits within fetched fasta files of peaks:
peak_sequences <- list.files(path = paste0(INPUTS, "FASTA/"), "*fasta")

hit_vec <- c()

# Creating an empty dataframe:
tbx5_hits <- data.frame(matrix(ncol = 3, nrow = 0)) %>%
                setNames(., c("Sample", "Peak_count", "Tbx5_hits"))

# Declaring a function that calculates Tbx5 motif hits in each sample:
find_motif_hits <- function(sequences) {
    for (i in 1:length(sequences)) {
        hits <- countPWM(as.matrix(mpwm), sequences[[i]], min.score = "75%")
        if (hits == 0) { next }
        else { hit_vec <- c(hit_vec, hits)}
    }
    return(sum(hit_vec))
}

# Declaring a function that calculates total peak count in each sample:
calculate_peaks <- function(filename) {
    name <- paste0(tools::file_path_sans_ext(filename), ".bb")
    file <- import(paste0(INPUTS, "BigBed/", name)) %>%
                    filter(seqnames %in% chr_abr)
    region_count <- length(file)
    return(region_count)
}

# Creating a new dataframe that stores data:
#   - generated sample name (used in plots);
#   - Tbx5 motif hit count;
#   - Total peak count;
#   - Tbx5 motif percentage.

# TEMPORARY COMMENTED CHUNK
# columns <- c("Sample", "Motif_count", "Peak_count", "Percentage") 
# tbx5_motifs <- data.frame(matrix(nrow = 0, ncol = length(columns))) 
 
# colnames(tbx5_motifs) <- columns

# for (i in 1:length(peak_sequences)) {
#     filename <- peak_sequences[i]
#     bb_file <- paste0(tools::file_path_sans_ext(filename), ".bb")
#     samples_plot_name <- samples$Graph_names[samples$Filename == bb_file]

#     seq <- readDNAStringSet(paste0(pth_fasta, filename))
#     motif <- find_motif_hits(seq)
#     peaks <- calculate_peaks(filename)
#     percentage <- round((motif / peaks) * 100, 2)

#     data_row <- c(samples_plot_name, motif, peaks, paste0(percentage, "%"))
#     tbx5_motifs[nrow(tbx5_motifs) + 1, ] <- data_row
# }

# write.csv(tbx5_motif, paste0(INTERMEDIATE_FILES, "tbx5_motif_data.csv"))
# THE END OF THE COMMENT

# Reading a file that stores information about Tbx5 motif counts and
# peak percentages:
motif_data <- read.csv(file = paste0(INTERMEDIATE_FILES, "tbx5_motif_data.csv"))

# Calling factor() function in order to maintain certain Sample order:
motif_data$Sample <- factor(motif_data$Sample, levels = motif_data$Sample)

# Subseting data to extract all columns except for 'X' column (column 1):
subset_df <- motif_data[, 2:5]

# 'Melting' the dataframe:
melted_df <- melt(subset_df)

# Creating a barplot that visualizes Tbx5 motif distribution within samples:
plot4 <- ggplot(data = melted_df, aes(x = Sample, y = value,
                                      fill = variable, label = value)) +
            geom_bar(stat = "identity", colour = "#35170450", size = 0.5,
                     width = 0.8) +
            scale_fill_manual(values = c("#e3a15e", "#c7633b"),
                              labels = c("Tbx5 motyvų skaičius",
                                         "Pikų skaičius")) +
            scale_y_continuous(labels = label_number(suffix = " K",
                                                     scale = 1e-3)) +
            guides(fill = guide_legend(title = "Spalvų paaiškinimas",
                                       size = 6)) +
            labs(title = "", x = "", y = "TF/Pikų skaičius") +
            theme(axis.text = element_text(size = 10, colour = "black"),
                  axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1,
                                             size = 12, face = "bold"),
                  axis.text.y = element_text(size = 12, face = "bold"),
                  axis.title.x = element_text(size = 14, colour = "black"),
                  axis.title.y = element_text(size = 14, colour = "black"),
                  panel.grid.major = element_line(color = "#eeeeee"),
                  plot.title = element_text(hjust = 0.5, size = 16,
                                            face = "bold"),
                  panel.background = element_rect(fill = "#eeeef1",
                                                  colour = "#4c0001"),
                  panel.grid.major.y = element_line(colour = "#cab5b5",
                                            size = 0.3, linetype = "dashed"),
                  panel.grid.minor.y = element_line(colour = "#cab5b5",
                                            size = 0.3, linetype = "dashed"),
                  panel.grid.major.x = element_line(colour = "#cab5b5",
                                            linetype = "longdash", size = 0.2),
                  panel.grid.minor.x = element_line(colour = "#cab5b5",
                                            linetype = "longdash", size = 0.2),
                  legend.position = c(0.79, 0.9),
                  legend.title = element_text(size = 12),
                  legend.text = element_text(size = 11))

# Saving generated image:
png(file = paste0(FIGURES, "tf_hit_percentage.png"), width = 400)
plot4
dev.off()
plot4
```

The bar chart below visualizes total identified motif count
for each sample.

```{r De_novo_motifs, fig.width = 5, fig.height = 7, fig.align = "center"}
# Reading a file that stores lists of motifs that were identified in
# each sample:
identified_motifs <- read.csv(file = paste0(INTERMEDIATE_FILES,
                              "identified_motifs.csv"))

# Getting rid of 'NA' values:
motif_counts <- data.frame(Motif_count = colSums(!is.na(identified_motifs)))

# Setting rownames to numbers:
rownames(motif_counts) <- NULL

# Making a new column with sample names:
motif_counts['Sample'] <- colnames(identified_motifs)

# Changing column order:
motif_counts <- motif_counts[, c(2, 1)]

# Calling factor() function in order to maintain certain Sample order:
motif_counts$Sample <- factor(motif_counts$Sample, levels = motif_counts$Sample)

# Creating a barplot that visualizes identified motif counts for each sample:
plot5 <- ggplot(motif_counts, aes(x = Sample, y = Motif_count)) +
            geom_bar(stat = "identity", fill = "#6b1313", width = .7,
                     color = "#4a1004") +
            labs(title = "", x = "", y = "Nustatytų motyvų skaičius") +
            ylim(0, 400) +
            geom_text(aes(label = Motif_count), color = "#030101", size = 5,
                          vjust = -1) +
            theme(axis.text = element_text(size = 10, colour = "black"),
                  axis.text.x = element_text(angle = 45, hjust = 1, size = 12,
                                             vjust = 1, face = "bold"),
                  axis.text.y = element_text(size = 12, face = "bold"),
                  axis.title.x = element_text(size = 14, colour = "black"),
                  axis.title.y = element_text(size = 14, colour = "black"),
                  panel.grid.major = element_line(color = "#eeeeee"),
                  plot.title = element_text(hjust = 0.5, size = 16,
                                            face = "bold"),
                  panel.background = element_rect(fill = "#eeeef1",
                                                  colour = "#3a1010"),
                  panel.grid.major.y = element_line(colour = "#cab5b5",
                                           size = 0.3, linetype = "dashed"),
                  panel.grid.minor.y = element_line(colour = "#cab5b5",
                                           size = 0.3, linetype = "dashed"),
                  panel.grid.major.x = element_line(colour = "#cab5b5",
                                           size = 0.2, linetype = "longdash"),
                  panel.grid.minor.x = element_line(colour = "#cab5b5",
                                           size = 0.2, linetype = "longdash"),
                  legend.position = c(0.777, 0.788))

# Saving generated image:
png(file = paste0(FIGURES, "motifs_in_samples.png"), width = 500)
plot5
dev.off()
plot5
```

Summarizing the results of the analysis it can be determined that Tbx5
transcription factor motif hit counts notably vary between the samples
where *Mus musculus* heart cells were affected with different treatments.\

On the other hand, it must be taken into account that it is difficult to
perform ChIP sequencing without making any mistakes and easily choosing
proper parameters. Due to this reason possible ChIP-seq aberrations
should be carefully considered.